stdlib/list.ml

```diff
@@ -125,6 +119,31 @@ let rec fold_right f l accu =
     [] -> accu
   | a::l -> f a (fold_right f l accu)
 
+let rec append_aux i l1 l2 =
+  match l1 with
+    [] -> l2
+  | first :: rest ->
+    if i > max_recursion_threshold / 2 then
+       rev_append (rev l1) l2
+    else
+       first :: append_aux (i + 1) rest l2
+
+let rev_flatten l =
+  fold_left (fun accu l -> rev_append l accu) [] l
+
+let rec flatten_aux i l =
+  if i > max_recursion_threshold / 2 then
+    rev (rev_flatten l)
+  else
+    match l with
+      [] -> []
+    | l1 :: r ->
+      append_aux 0 l1 (flatten_aux (i + 1) r)
+
```

Comment:
Why do you split `max_recursion_threshold` ?
It seems to me that you should rather call `append_aux i l1 (flatten_aux (i + 1) r)`, since this is the same stack.

Of course the trouble with this whole approach is that to be extensible we would have to export `max_recusion_threshold` and the `_aux` version.